wardrobe/tradable_kitties/src/lib.rs

```diff
@@ -0,0 +1,446 @@
+        { TradeableKittyError::NumberOfInputOutputMismatch }
+    );
+
+    let mut dna_to_tdkitty_set: BTreeSet<KittyDNA> = BTreeSet::new();
+    let mut dna_to_kitty_set: BTreeSet<KittyDNA> = BTreeSet::new();
+
+    for i in 0..kitty_data.len() {
+        let utxo_kitty = kitty_data[i]
+            .extract::<KittyData>()
+            .map_err(|_| TradeableKittyError::BadlyTyped)?;
+
+        if dna_to_kitty_set.contains(&utxo_kitty.dna) {
+            return Err(TradeableKittyError::DuplicateKittyFound);
+        } else {
+            dna_to_kitty_set.insert(utxo_kitty.clone().dna);
+        }
+
+        let utxo_tradable_kitty = tradable_kitty_data[i]
+            .extract::<TradableKittyData>()
+            .map_err(|_| TradeableKittyError::BadlyTyped)?;
+
+        if dna_to_tdkitty_set.contains(&utxo_tradable_kitty.kitty_basic_data.dna) {
+            return Err(TradeableKittyError::DuplicateTradableKittyFound);
+        } else {
+            dna_to_tdkitty_set.insert(utxo_tradable_kitty.clone().kitty_basic_data.dna);
+        }
+
+        ensure!(
+            utxo_kitty == utxo_tradable_kitty.kitty_basic_data,
+            TradeableKittyError::KittyBasicPropertiesAltered
+        );
+        ensure!(
+            utxo_tradable_kitty.price != None, // basic kitty data is unaltered
+            TradeableKittyError::KittyPriceCantBeNone  // this need to be chan
+        );
+    }
+
+    return Ok(0);
+}
+
+impl<const ID: u8> SimpleConstraintChecker for TradableKittyConstraintChecker<ID> {
+    type Error = TradeableKittyError;
+
+    fn check(
+        &self,
+        input_data: &[DynamicallyTypedData],
+        _peeks: &[DynamicallyTypedData],
+        output_data: &[DynamicallyTypedData],
+    ) -> Result<TransactionPriority, Self::Error> {
+        match &self {
+            Self::ListKittyForSale => {
+                check_can_list_kitty_for_sale(&input_data, &output_data)?;
+                return Ok(0);
+            }
+            Self::DelistKittyFromSale => {
+                check_can_delist_kitty_from_sale(&input_data, &output_data)?;
+            }
+            Self::UpdateKittyPrice => {
+                check_kitty_price_update(input_data, output_data)?;
+            }
+            Self::UpdateKittyName => {
+                let mut input_basic_kitty_data: Vec<DynamicallyTypedData> = Vec::new();
+                let mut output_basic_kitty_data: Vec<DynamicallyTypedData> = Vec::new();
+                let _ = extract_basic_kitty_list(&input_data, &mut input_basic_kitty_data)?;
+                let _ = extract_basic_kitty_list(&output_data, &mut output_basic_kitty_data)?;
+                kitties::can_kitty_name_be_updated(
+                    &input_basic_kitty_data,
+                    &output_basic_kitty_data,
+                )?;
+            }
+            Self::Buy => {
+                check_can_buy::<ID>(input_data, output_data)?;
+                return Ok(0);
```

Comment:
Same with the early return here. If you are returning different values, it could make sense, but in this case, the final `Ok(0)` will be just fine.

Reply:
fixed.